vendor/github.com/hashicorp/go-raftchunking/fsm.go

```diff
@@ -66,74 +56,59 @@ func (c *ChunkingFSM) Apply(l *raft.Log) interface{} {
 		return c.underlying.Apply(l)
 	}
 
+	if l.Term != c.lastTerm {
+		// Term has changed. A raft library client that was applying chunks
+		// should get an error that it's no longer the leader and bail, and
```

Comment:
What returns that error? I'm not _sure_ that there aren't cases where the raft term will change without the leader loosing it's lease and stepping down fully and so completing all inflight ops with NO_LEADER which is the current case we retry (in client). That _might_ be true but I'm not sure it is. Even if it's intended to be true I'm not sure it can be relied on with all possible message timings/interleavings etc.

We can look and think some more but it's not obvious to me this is definitely correct. I don't _think_ it can cause corruption since all replicas must agree on the term for each log so they will all behave the same, but it _may_ be possible for a client not to see an error during chunk writing even though term changed after chunk 1 and the state was discarded and then it continues to write lots more data chunks only to find at the end it can't apply because the first one is now missing. At worst that's wasted effort and should be rare but I'd like to understand the guarantees around this better.

Reply:
Hm. See previous comment; Brian seemed to think this wouldn't be a problem, but I'm not really sure. I assume if the client thinks it applied all chunks what would happen is that one of its futures would simply never return / time out. Which isn't great for the client, but won't cause corruption.

If you agree that this shouldn't cause corruption, maybe what I can do for now is write the term into each chunk info but leave the logic as is. That way we can change the logic later if we want but will have the necessary metadata still there in existing snapshots and state just in case we decide we want it. Thoughts?

Reply:
I'm also happy to just put that error back. It's easy to check that the term didn't change. I'll do that.

Reply:
Actually I remember now why I didn't do that. Let's say we are clearing the map on term change, which we are, because that's how we're GCing. Now let's say that we see an entry with a new term. We can reject _that log_ with an ErrTermChanged, but unless we keep all of the old ops around anyways instead of clearing storage, we can't know if the _next log_ will also need that rejection. So if storage is cleared, that next log coming in just looks like the first log for a given operation, which we accept and hold. In other words -- to the client the behavior will be the same.

I actually wonder if there's something sneaky here (and bear with me, as I think I also have a solution). Let's say a client writes 11 out of 12 logs on one term. Those futures resolve to nil error. Then for some reason the 12th log even though it has a new term gets applied, so we clear the existing map and store that as the only log for that opnum. Now the client thinks that all operations completed successfully when in fact it was never Applied to the underlying store.

The only way this wouldn't be a problem is if the client can know that the last future should return something specific that isn't nil, because the chunking FSM will always return nil on success except for the last Apply. If we can guarantee that, then we can actually have the chunking apply command in go-raftchunking return an error if the last future returns nil (or conversely, have it not do anything itself, and have the client using it know that it needs to expect a non-nil response and retry if not).

In Consul Apply will always return non-nil unless there is an unknown message type. I wonder if instead of returning nil for unknown message type we can return a specific error. Then we guarantee that the final apply should not be nil and the error can be handled appropriately (even though in practice it just means discarding it).

Reply:
I see. I think making `nil` be a very special case for chunking is kinda gross - it feels like something it would be very easy to forget about and inadvertently mess up later for one but even worse `ErrSuccess` is a pretty bad code smell that indicates hackery at play to me!

How about we do exactly the same, but we make the Successful return value of a final chunk apply always be a specific wrapper value like `ChunkedSuccess{result interface{}}`.

That way the chunk apply code can disambiguate between a last chunk that was actually applied and one that was just nil because we didn't think it was the last chunk, and the client can unwrap to get the real result.

This makes no assumptions about anything else in the FSM not returning nil etc.

Reply:
Sounds good.